harness/determined/layers/_workload_sequencer.py

```diff
@@ -0,0 +1,507 @@
+            self.state.last_val = self.state.total_batches
+
+    def as_batches(
+        self,
+        batches: Optional[int] = None,
+        records: Optional[int] = None,
+        epochs: Optional[int] = None,
+    ) -> int:
+        if sum((batches is not None, records is not None, epochs is not None)) != 1:
+            raise ValueError(f"invalid length: batches={batches} records={records} epochs={epochs}")
+        if batches is not None:
+            return batches
+        if records is not None:
+            check.gt(self.global_batch_size, 0, "global_batch_size must be positive")
+            return max(records // self.global_batch_size, 1)
+        if epochs is not None:
+            check.is_instance(self.records_per_epoch, int, "length must be an integer")
+            assert self.records_per_epoch is not None
+            check.gt(self.global_batch_size, 0, "global_batch_size must be positive")
+            return max((epochs * self.records_per_epoch) // self.global_batch_size, 1)
+        # Make mypy happy.
+        raise ValueError("invalid length")
+
+    def check_for_preemption(self) -> None:
+        assert self.preemption is not None
+        if self.preemption.should_preempt(chief_only=True):
+            raise ShouldExit()
+
+    def train(self, num_batches: int, op: _searcher.SearcherOp) -> WorkloadGenerator:
+        # Report a train step is starting.
+        self.training.set_status("training")
+
+        wkld = workload.Workload(
+            kind=workload.Workload.Kind.RUN_STEP,
+            e_id=self._exp_id,
+            t_id=self._trial_id,
+            s_id=self.state.step_id + 1,
+            num_batches=num_batches,
+            total_batches_processed=self.state.total_batches,
+        )
+
+        response = yield from yield_and_await_response(wkld)
+
+        # Train step is complete, process the result.
+
+        if isinstance(response, workload.InvalidHP):
+            # Exit before reporting metrics (which would be empty anyway).
+            self.training.report_early_exit(_training.EarlyExitReason.INVALID_HP)
+            raise ShouldExit()
+
+        metrics = response.get("metrics", {}).get("avg_metrics", {})
+        batch_metrics = response.get("metrics", {}).get("batch_metrics", [])
+        num_inputs = response.get("metrics", {}).get("num_inputs")
+
+        self.state.total_batches += num_batches
+        if num_inputs is None or self.state.total_records is None:
+            self.state.total_records = None
+        else:
+            self.state.total_records += num_inputs
+        self.state.step_id += 1
+        self.training.report_training_metrics(
+            total_batches=self.state.total_batches,
+            metrics=metrics,
+            batch_metrics=batch_metrics,
+            total_records=self.state.total_records,
+        )
+
+        # Report progress to the searcher.  For historical reasons we only deal in batches.
```

Comment:
The WorkloadSequencer does the same crappy batch math that the master-side one did.  I don't want to torture TFKeras users with another twist of complexity; I'd rather just adopt the newer, simpler behavior in the future high-level keras API.

This logic here is basically:
- if the searcher is in batches, report `total_batches`
- if the searcher is in records, report `total_batches * records_per_batch`, ignoring how many records have actually been trained
- if the searcher is in epochs, report `total_batches / (target_epochs * batches_per_epoch)`, ignoring how many epochs the dataset has actually done

So the correct unit is always used in the REST API, but for epochs and records the correct unit is calculated from `total_batches` rather than measured directly.

Reply:
oh ok yeah, misread. lgtm.